timeinterval/timeinterval.go

```diff
@@ -0,0 +1,473 @@
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// TimeInterval describes intervals of time. ContainsTime will tell you if a golang time is contained
+// within the interval.
+type TimeInterval struct {
+	Times       []TimeRange       `yaml:"times,omitempty"`
+	Weekdays    []WeekdayRange    `yaml:"weekdays,flow,omitempty"`
+	DaysOfMonth []DayOfMonthRange `yaml:"days_of_month,flow,omitempty"`
+	Months      []MonthRange      `yaml:"months,flow,omitempty"`
+	Years       []YearRange       `yaml:"years,flow,omitempty"`
+}
+
+/* TimeRange represents a range of minutes within a 1440 minute day, exclusive of the End minute. A day consists of 1440 minutes.
+   For example, 5:00PM to End of the day would Begin at 1020 and End at 1440. */
+type TimeRange struct {
+	StartMinute int
+	EndMinute   int
+}
+
+// InclusiveRange is used to hold the Beginning and End values of many time interval components
+type InclusiveRange struct {
+	Begin int
+	End   int
+}
+
+// A WeekdayRange is an inclusive range between [0, 6] where 0 = Sunday
+type WeekdayRange struct {
+	InclusiveRange
+}
+
+// A DayOfMonthRange is an inclusive range that may have negative Beginning/End values that represent distance from the End of the month Beginning at -1
+type DayOfMonthRange struct {
+	InclusiveRange
+}
+
+// A MonthRange is an inclusive range between [1, 12] where 1 = January
+type MonthRange struct {
+	InclusiveRange
+}
+
+// A YearRange is a positive inclusive range
+type YearRange struct {
+	InclusiveRange
+}
+
+type yamlTimeRange struct {
+	StartTime string `yaml:"start_time"`
+	EndTime   string `yaml:"end_time"`
+}
+
+// A range with a Beginning and End that can be represented as strings
+type stringableRange interface {
+	setBegin(int)
+	setEnd(int)
+	// Try to map a member of the range into an integer.
+	memberFromString(string) (int, error)
+}
+
+func (ir *InclusiveRange) setBegin(n int) {
+	ir.Begin = n
+}
+
+func (ir *InclusiveRange) setEnd(n int) {
+	ir.End = n
+}
+
+func (ir *InclusiveRange) memberFromString(in string) (out int, err error) {
+	out, err = strconv.Atoi(in)
+	if err != nil {
+		return -1, err
+	}
+	return out, nil
+}
+
+func (r *WeekdayRange) memberFromString(in string) (out int, err error) {
+	out, ok := daysOfWeek[in]
+	if !ok {
+		return -1, fmt.Errorf("%s is not a valid weekday", in)
+	}
+	return out, nil
+}
+
+func (r *MonthRange) memberFromString(in string) (out int, err error) {
+	out, ok := months[in]
+	if !ok {
+		out, err = strconv.Atoi(in)
+		if err != nil {
+			return -1, fmt.Errorf("%s is not a valid month", in)
+		}
+	}
+	return out, nil
+}
+
+var daysOfWeek = map[string]int{
+	"sunday":    0,
+	"monday":    1,
+	"tuesday":   2,
+	"wednesday": 3,
+	"thursday":  4,
+	"friday":    5,
+	"saturday":  6,
+}
+var daysOfWeekInv = map[int]string{
+	0: "sunday",
+	1: "monday",
+	2: "tuesday",
+	3: "wednesday",
+	4: "thursday",
+	5: "friday",
+	6: "saturday",
+}
+
+var months = map[string]int{
+	"january":   1,
```

Comment:
Months in English start with a capital letter if I'm right

Reply:
They're just stored as lowercase here for convenience and then downcased during the lookup, so users can specify it with or without capitals.

Reply:
Ok, then this is not tested

Reply:
Thanks, I'll add tests for this.